Add support for Web Extension windows APIs.

xeenon opened the pull request:
#### 589a9da68d4ea19dffeee69c65375b872cca7c1b
<pre>
Add support for Web Extension windows APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260155">https://bugs.webkit.org/show_bug.cgi?id=260155</a>
rdar://problem/114736173

Reviewed by Brian Weinstein.

Fully plumbs the windows API from JavaScript up to the Objective-C API.

* Source/WebKit/DerivedSources-input.xcfilelist: Addeed new files.
* Source/WebKit/DerivedSources.make: Added new files.
* Source/WebKit/Scripts/webkit/messages.py: Added new extension types.
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp: Updated with make command mentioned in messages.py comment.
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
(WebKit::isNone): Added.
(WebKit::isValid): Added.
(WebKit::toWebExtensionTabIdentifier): Added.
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabParameters.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h.
* Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
(WebKit::isNone): Added.
(WebKit::isCurrent): Added.
(WebKit::isValid): Added.
(WebKit::toWebExtensionWindowIdentifier): Added.
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowParameters.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h.
* Source/WebKit/Shared/Extensions/_WKWebExtensionUtilities.mm:
(classToClassString): Use dot-syntax.
(+[_WKWebExtensionUtilities validateContentsOfDictionary:requiredKeys:optionalKeys:keyToExpectedValueType:outExceptionString:]):
Added support for NSSet input, where any class or array in the set can match. This is needed for the windows API that allows a single
URL or array of URLs. Also removed error logging, since this can be used for partial validation and it was causing log noise.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptionsInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm: Added.
(WebKit::matchesFilter): Added.
(WebKit::WebExtensionContext::windowsCreate): Added.
(WebKit::WebExtensionContext::windowsGet): Added.
(WebKit::WebExtensionContext::windowsGetLastFocused): Added.
(WebKit::WebExtensionContext::windowsGetAll): Added.
(WebKit::WebExtensionContext::windowsUpdate): Added.
(WebKit::WebExtensionContext::windowsRemove): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getOrCreateWindow): Tweaked logging.
(WebKit::WebExtensionContext::getWindow): Added current window support with FIXME.
(WebKit::WebExtensionContext::getOrCreateTab): Tweaked logging.
(WebKit::WebExtensionContext::getTab): Ditto.
(WebKit::WebExtensionContext::frontmostWindow): Added.
(WebKit::WebExtensionContext::didOpenWindow): Added logging.
(WebKit::WebExtensionContext::didCloseWindow): Ditto.
(WebKit::WebExtensionContext::didFocusWindow): Ditto.
(WebKit::WebExtensionContext::didOpenTab): Ditto.
(WebKit::WebExtensionContext::didCloseTab): Ditto.
(WebKit::WebExtensionContext::didActivateTab): Ditto.
(WebKit::WebExtensionContext::didSelectTabs): Ditto.
(WebKit::WebExtensionContext::didMoveTab): Ditto.
(WebKit::WebExtensionContext::didReplaceTab): Ditto.
(WebKit::WebExtensionContext::didChangeTabProperties): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::extensionContext const): Added here to fix build.
(WebKit::WebExtensionTab::parameters const): Added.
(WebKit::WebExtensionTab::toggleReaderMode): Call completion handler.
(WebKit::WebExtensionTab::mute): Ditto.
(WebKit::WebExtensionTab::unmute): Ditto.
(WebKit::WebExtensionTab::setZoomFactor): Ditto.
(WebKit::WebExtensionTab::loadURL): Ditto.
(WebKit::WebExtensionTab::reload): Ditto.
(WebKit::WebExtensionTab::reloadFromOrigin): Ditto.
(WebKit::WebExtensionTab::goBack): Ditto.
(WebKit::WebExtensionTab::goForward): Ditto.
(WebKit::WebExtensionTab::activate): Ditto.
(WebKit::WebExtensionTab::select): Ditto.
(WebKit::WebExtensionTab::duplicate): Ditto.
(WebKit::WebExtensionTab::close): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm:
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::WebExtensionWindow::extensionContext const): Added here to fix build.
(WebKit::WebExtensionWindow::parameters const): Added.
(WebKit::toAPI): Added.
(WebKit::WebExtensionWindow::setState): Call completion handler.
(WebKit::WebExtensionWindow::focus):
(WebKit::WebExtensionWindow::normalizedFrame const): Added.
(WebKit::WebExtensionWindow::frame const): Added rect standardization.
(WebKit::WebExtensionWindow::setFrame): Call completion handler.
(WebKit::WebExtensionWindow::screenFrame const): Added.
(WebKit::WebExtensionWindow::close): Call completion handler.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::extensionContext const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
(WebKit::WebExtensionWindow::extensionContext const): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI): Added stub.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::toWebAPI): Added.
(WebKit::toTypeImpl): Added.
(WebKit::toStateImpl): Added.
(WebKit::WebExtensionAPIWindows::parsePopulateTabs): Added.
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter): Added.
(WebKit::WebExtensionAPIWindows::parseWindowGetOptions): Added.
(WebKit::WebExtensionAPIWindows::parseWindowCreateOptions): Added.
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Added.
(WebKit::WebExtensionAPIWindows::isPropertyAllowed): Added implementation.
(WebKit::WebExtensionAPIWindows::createWindow): Added implementation.
(WebKit::WebExtensionAPIWindows::get): Added implementation.
(WebKit::WebExtensionAPIWindows::getCurrent): Added implementation.
(WebKit::WebExtensionAPIWindows::getLastFocused): Added implementation.
(WebKit::WebExtensionAPIWindows::getAll): Added implementation.
(WebKit::WebExtensionAPIWindows::update): Added implementation.
(WebKit::WebExtensionAPIWindows::remove): Added implementation.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument): Use WebExtensionAPIWindows filter.
(WebKit::WebExtensionAPIWindowsEvent::addListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindows.h:
(WebKit::WebExtensionAPIWindows::windowIdentifierNone const): Added.
(WebKit::WebExtensionAPIWindows::windowIdentifierCurrent const): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindowsEvent.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindows.idl: Added NeedsPage on a couple methods.
* Tools/TestWebKitAPI/SourcesCocoa.txt: Added new files.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionTab.mm:
(TestWebKitAPI::TEST): Use new TestWebExtensionWindow object.
(-[TestWindowForTabs tabsForWebExtensionContext:]): Deleted.
(-[TestWindowForTabs activeTabForWebExtensionContext:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm:
(TestWebKitAPI::TEST): Use new TestWebExtensionWindow object.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openNewWindowWithOptions:forExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionsDelegate webExtensionController:openNewTabWithOptions:forExtensionContext:completionHandler:]): Added.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionWindow init]): Added.
(-[TestWebExtensionWindow tabsForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow activeTabForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow windowTypeForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow windowStateForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow setWindowState:forWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow isUsingPrivateBrowsingForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow screenFrameForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow frameForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow setFrame:forWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow focusForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow closeForWebExtensionContext:completionHandler:]): Added.

Canonical link: <a href="https://commits.webkit.org/267522@main">https://commits.webkit.org/267522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4477d2ac818a54d24b5bfbc55994db71bf28371f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16837 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17160 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18618 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15772 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17298 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17036 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/17399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/14579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19419 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/16966 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15264 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/15632 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/15431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19747 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16037 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15210 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19574 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2075 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in:
```diff
@@ -0,0 +1,59 @@
+    None,
+    Normal,
+    Popup,
+    All,
```
What does the `All` type mean?

xeenon replied:
It makes more sense in the header, but it means `Normal | Popup`, which is used as a filter for some of the window APIs.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1155,21 +1155,28 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (!isValid(identifier))
+        return nullptr;
+
+    // FIXME: Use the page identifier to get the current window based on web view.
```
which bug is tracking this?

xeenon replied:
I'll file one and add it.

b-weinstein commented on Source/WebKit/DerivedSources.make:
```diff
@@ -494,14 +494,16 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	Shared/Gamepad/GamepadData.serialization.in \
 	Shared/GPUProcessConnectionParameters.serialization.in \
-        Shared/GoToBackForwardItemParameters.serialization.in \
+	Shared/GoToBackForwardItemParameters.serialization.in \
```
It looks like the whitespace changed? And it appears to be more consistent with the rest of the file.

xeenon replied:
It did, it was a tab before I think.

b-weinstein commented on Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in:
```diff
@@ -0,0 +1,59 @@
+    std::optional<bool> privateBrowsing;
+};
+
+[Nested] enum class WebKit::WebExtensionWindow::PopulateTabs : bool;
```
Does ` : bool` mean we will get PopulateTabs::Yes and PopulateTabs::No?

xeenon replied:
Yeah, the `: bool` is just the storage type.

b-weinstein commented on Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
```diff
@@ -25,11 +25,82 @@
+
+inline bool isValid(std::optional<WebExtensionWindowIdentifier> identifier)
+{
+    return identifier && !isNone(identifier.value());
```
is `current` considered `valid`? I'm guessing so

xeenon replied:
Yeah

b-weinstein commented on Source/WebKit/Shared/Extensions/_WKWebExtensionUtilities.mm:
```diff
@@ -80,38 +81,88 @@ + (BOOL)validateContentsOfDictionary:(NSDictionary<NSString *, id> *)dictionary
+        Class expectedElementType = validClassesArray.firstObject;
+        NSArray *arrayValue = dynamic_objc_cast<NSArray>(value);
+        if (!arrayValue) {
+            errorString = [NSString stringWithFormat:@"Expected an array for '%@'.", key];
```
Nit: Do we want to put the class we found instead?

xeenon replied:
Yeah that would be good. Will do.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
```diff
@@ -125,10 +135,23 @@ WK_API_AVAILABLE(macos(13.3), ios(16.4))
  which should be provided if any errors occurred.
- @discussion No action is performed if not implemented.
+ @discussion On macOS, the implementation of both `frameForWebExtensionContext:` and
+ `screenFrameForWebExtensionContext:` is a prerequisite. On iOS, only `frameForWebExtensionContext:`
```
Nit: `are prerequisites`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm:
```diff
@@ -0,0 +1,296 @@
+#if PLATFORM(MAC)
+        // Window coordinates on macOS have the origin in the bottom-left corner.
+        // Web Extensions have window coordinates in the top-left corner.
+        CGRect screenFrame = NSScreen.mainScreen.frame;
```
Will all windows be created on the main screen? Or would this be wrong if the browser was in the non-main screen?

xeenon replied:
On macOS the other screens are rooted in the main screen coords. So an extension could still position on a second screen by using negative positions or positive positions wider than the main screen, depending on where the other screen is positioned. So using mainScreen here is really the only choice.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -80,11 +80,23 @@
 
+WebExtensionTabParameters WebExtensionTab::parameters() const
+{
+    // FIXME: Implement.
```
Nit: Radar for this?

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -38,12 +38,20 @@
 
+NSDictionary *toWebAPI(const WebExtensionTabParameters& parameters)
+{
+    // FIXME: Implemnet.
```
Bug # + a typo here (Implement)

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
```diff
@@ -35,69 +35,501 @@
+static NSString *alwaysOnTopKey = @"alwaysOnTop";
+
+static NSString *urlKey = @"url";
+static NSString *tabIdKey = @"tabId";
```
Should these be `const`? Or does that not really matter here?

xeenon replied:
Yeah they should be!

xeenon replied:
That ended up causing build issues, since all uses need to be const and it was annoying.

xeenon committed:
Add support for Web Extension windows APIs.
https://bugs.webkit.org/show_bug.cgi?id=260155
rdar://problem/114736173

Reviewed by Brian Weinstein.

Fully plumbs the windows API from JavaScript up to the Objective-C API.

* Source/WebKit/DerivedSources-input.xcfilelist: Addeed new files.
* Source/WebKit/DerivedSources.make: Added new files.
* Source/WebKit/Scripts/webkit/messages.py: Added new extension types.
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp: Updated with make command mentioned in messages.py comment.
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h:
(WebKit::isNone): Added.
(WebKit::isValid): Added.
(WebKit::toWebExtensionTabIdentifier): Added.
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionTabParameters.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h.
* Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h:
(WebKit::isNone): Added.
(WebKit::isCurrent): Added.
(WebKit::isValid): Added.
(WebKit::toWebExtensionWindowIdentifier): Added.
(WebKit::toWebAPI): Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowParameters.h: Copied from Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h.
* Source/WebKit/Shared/Extensions/_WKWebExtensionUtilities.mm:
(classToClassString): Use dot-syntax.
(+[_WKWebExtensionUtilities validateContentsOfDictionary:requiredKeys:optionalKeys:keyToExpectedValueType:outExceptionString:]):
Added support for NSSet input, where any class or array in the set can match. This is needed for the windows API that allows a single
URL or array of URLs. Also removed error logging, since this can be used for partial validation and it was causing log noise.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptionsInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm: Added.
(WebKit::matchesFilter): Added.
(WebKit::WebExtensionContext::windowsCreate): Added.
(WebKit::WebExtensionContext::windowsGet): Added.
(WebKit::WebExtensionContext::windowsGetLastFocused): Added.
(WebKit::WebExtensionContext::windowsGetAll): Added.
(WebKit::WebExtensionContext::windowsUpdate): Added.
(WebKit::WebExtensionContext::windowsRemove): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getOrCreateWindow): Tweaked logging.
(WebKit::WebExtensionContext::getWindow): Added current window support with FIXME.
(WebKit::WebExtensionContext::getOrCreateTab): Tweaked logging.
(WebKit::WebExtensionContext::getTab): Ditto.
(WebKit::WebExtensionContext::frontmostWindow): Added.
(WebKit::WebExtensionContext::didOpenWindow): Added logging.
(WebKit::WebExtensionContext::didCloseWindow): Ditto.
(WebKit::WebExtensionContext::didFocusWindow): Ditto.
(WebKit::WebExtensionContext::didOpenTab): Ditto.
(WebKit::WebExtensionContext::didCloseTab): Ditto.
(WebKit::WebExtensionContext::didActivateTab): Ditto.
(WebKit::WebExtensionContext::didSelectTabs): Ditto.
(WebKit::WebExtensionContext::didMoveTab): Ditto.
(WebKit::WebExtensionContext::didReplaceTab): Ditto.
(WebKit::WebExtensionContext::didChangeTabProperties): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::extensionContext const): Added here to fix build.
(WebKit::WebExtensionTab::parameters const): Added.
(WebKit::WebExtensionTab::toggleReaderMode): Call completion handler.
(WebKit::WebExtensionTab::mute): Ditto.
(WebKit::WebExtensionTab::unmute): Ditto.
(WebKit::WebExtensionTab::setZoomFactor): Ditto.
(WebKit::WebExtensionTab::loadURL): Ditto.
(WebKit::WebExtensionTab::reload): Ditto.
(WebKit::WebExtensionTab::reloadFromOrigin): Ditto.
(WebKit::WebExtensionTab::goBack): Ditto.
(WebKit::WebExtensionTab::goForward): Ditto.
(WebKit::WebExtensionTab::activate): Ditto.
(WebKit::WebExtensionTab::select): Ditto.
(WebKit::WebExtensionTab::duplicate): Ditto.
(WebKit::WebExtensionTab::close): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm:
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::WebExtensionWindow::extensionContext const): Added here to fix build.
(WebKit::WebExtensionWindow::parameters const): Added.
(WebKit::toAPI): Added.
(WebKit::WebExtensionWindow::setState): Call completion handler.
(WebKit::WebExtensionWindow::focus):
(WebKit::WebExtensionWindow::normalizedFrame const): Added.
(WebKit::WebExtensionWindow::frame const): Added rect standardization.
(WebKit::WebExtensionWindow::setFrame): Call completion handler.
(WebKit::WebExtensionWindow::screenFrame const): Added.
(WebKit::WebExtensionWindow::close): Call completion handler.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::extensionContext const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
(WebKit::WebExtensionWindow::extensionContext const): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI): Added stub.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::toWebAPI): Added.
(WebKit::toTypeImpl): Added.
(WebKit::toStateImpl): Added.
(WebKit::WebExtensionAPIWindows::parsePopulateTabs): Added.
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter): Added.
(WebKit::WebExtensionAPIWindows::parseWindowGetOptions): Added.
(WebKit::WebExtensionAPIWindows::parseWindowCreateOptions): Added.
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Added.
(WebKit::WebExtensionAPIWindows::isPropertyAllowed): Added implementation.
(WebKit::WebExtensionAPIWindows::createWindow): Added implementation.
(WebKit::WebExtensionAPIWindows::get): Added implementation.
(WebKit::WebExtensionAPIWindows::getCurrent): Added implementation.
(WebKit::WebExtensionAPIWindows::getLastFocused): Added implementation.
(WebKit::WebExtensionAPIWindows::getAll): Added implementation.
(WebKit::WebExtensionAPIWindows::update): Added implementation.
(WebKit::WebExtensionAPIWindows::remove): Added implementation.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::invokeListenersWithArgument): Use WebExtensionAPIWindows filter.
(WebKit::WebExtensionAPIWindowsEvent::addListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindows.h:
(WebKit::WebExtensionAPIWindows::windowIdentifierNone const): Added.
(WebKit::WebExtensionAPIWindows::windowIdentifierCurrent const): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWindowsEvent.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWindows.idl: Added NeedsPage on a couple methods.
* Tools/TestWebKitAPI/SourcesCocoa.txt: Added new files.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionTab.mm:
(TestWebKitAPI::TEST): Use new TestWebExtensionWindow object.
(-[TestWindowForTabs tabsForWebExtensionContext:]): Deleted.
(-[TestWindowForTabs activeTabForWebExtensionContext:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm:
(TestWebKitAPI::TEST): Use new TestWebExtensionWindow object.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openNewWindowWithOptions:forExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionsDelegate webExtensionController:openNewTabWithOptions:forExtensionContext:completionHandler:]): Added.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionWindow init]): Added.
(-[TestWebExtensionWindow tabsForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow activeTabForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow windowTypeForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow windowStateForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow setWindowState:forWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow isUsingPrivateBrowsingForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow screenFrameForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow frameForWebExtensionContext:]): Added.
(-[TestWebExtensionWindow setFrame:forWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow focusForWebExtensionContext:completionHandler:]): Added.
(-[TestWebExtensionWindow closeForWebExtensionContext:completionHandler:]): Added.

Canonical link: https://commits.webkit.org/267522@main
